src/webapi/rendering_context.rs

```diff
@@ -18,6 +20,160 @@ pub trait RenderingContext {
+pub struct TextMetrics(Reference);
+
+/// The type of compositing operation to apply when drawing new shapes, where type is a string 
+/// identifying which of the twelve compositing operations to use.
+/// 
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/globalCompositeOperation)
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum CompositeOperation {
+    SourceOver,
+    SourceIn,
+    SourceOut,
+    SourceAtop,
+    DestinationOver,
+    DestinationIn,
+    DestinationAtop,
+    Lighter,
+    Copy,
+    Xor,
+    Multiply,
+    Screen,
+    Overlay,
+    Darken,
+    Lighten,
+    ColorDodge,
+    ColorBurn,
+    HardLight,
+    SoftLight,
+    Difference,
+    Exclusion,
+    Hue,
+    Saturation,
+    Color,
+    Luminosity
+}
+
+/// The algorithm by which to determine if a point is inside a path or outside a path.
+/// 
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/fill)
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum FillRule {
+    NonZero,
+    EvenOdd
+}
+
+/// How the end points of every line are drawn.
+/// 
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/lineCap)
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum LineCap {
+    Butt,
+    Round,
+    Square
+}
+
+/// determines how two connecting segments (of lines, arcs or curves) with non-zero lengths in a shape are 
+/// joined together (degenerate segments with zero lengths, whose specified endpoints and control points are 
+/// exactly at the same position, are skipped).
+/// 
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/lineJoin)
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum LineJoin {
+    Bevel,
+    Round,
+    Miter
+}
+
+/// A DOMString indicating how to repeat the image.
+/// 
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/createPattern)
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Repetition {
+    Repeat,
+    RepeatX,
+    RepeatY,
+    NoRepeat
+}
+
+/// Specifies text alignment
+/// 
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/textAlign)
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum TextAlign {
+    Left,
+    Right,
+    Center,
+    Start,
+    End
+}
+
+/// Text baseline being used when drawing text
+/// 
+/// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/textBaseline)
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum TextBaseline {
+    Top,
+    Hanging,
+    Middle,
+    Alphabetic,
+    Ideographic,
+    Bottom
+}
+
+error_enum_boilerplate! {
+    IndexSizeError,
+    SyntaxError
+}
+
+error_enum_boilerplate! {
+    NotSupportedError,
+    SyntaxError
+}
+
+error_enum_boilerplate! {
+    TypeMismatchError,
+    SyntaxError
+}
```

Comment:
These aren't correct from what I can see. (: 

The idea of the `error_enum_boilerplate!` is to define a localized error enum which encompasses all of the errors which can be thrown by the original API.

```
error_enum_boilerplace! {
    <NameOfTheLocalErrorEnum>,
    <Error1>, <Error2>, <Error3>, ...
```

Reply:
OHHHHHHHHHH I totally misunderstood that macro! I even tried (unknowingly) to re-implement it myself! Makes a lot more sense now :)